examples/form-fields.story.js

```diff
@@ -0,0 +1,310 @@
+      errors.inventory.fractions = true;
+  }
+
+  // Formik will think there are errors when the object returned as a
+  // validation result has at least one key. We therefore have to omit
+  // all empty fields inside the error object.
+  // When an empty object is returned, formik will assume there were no errors
+  return omitEmpty(errors);
+};
+
+class ProductForm extends React.Component {
+  static displayName = 'ProductForm';
+  static propTypes = {
+    formik: PropTypes.shape({
+      values: PropTypes.shape({
+        version: PropTypes.number.isRequired,
+        key: PropTypes.string.isRequired,
+        // Formik provides us with either a number or an empty string in case
+        // the value can not be parsed
+        inventory: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
+          .isRequired,
+      }).isRequired,
+      touched: PropTypes.shape({
+        key: PropTypes.bool,
+        inventory: PropTypes.bool,
+      }),
+      errors: PropTypes.shape({
+        key: PropTypes.shape({
+          missing: PropTypes.bool,
+          duplicate: PropTypes.bool,
+        }),
+        inventory: PropTypes.shape({
+          negative: PropTypes.bool,
+          fractions: PropTypes.bool,
+          missing: PropTypes.bool,
+        }),
+      }).isRequired,
+      dirty: PropTypes.bool.isRequired,
+      handleChange: PropTypes.func.isRequired,
+      handleBlur: PropTypes.func.isRequired,
+      handleSubmit: PropTypes.func.isRequired,
+      handleReset: PropTypes.func.isRequired,
+      isSubmitting: PropTypes.bool.isRequired,
+    }).isRequired,
+  };
+  render() {
+    return (
+      <Spacings.Stack scale="m">
+        <TextField
+          title="Key"
+          name="key"
+          value={this.props.formik.values.key}
+          onChange={this.props.formik.handleChange}
+          onBlur={this.props.formik.handleBlur}
+          isDisabled={this.props.formik.isSubmitting}
+          isTouched={this.props.formik.touched.key}
+          errors={this.props.formik.errors.key}
+          renderError={key => {
+            // This example shows how to handle custom errors on top of the
+            // predefined errors of FieldErrors which this component and other
+            // Field components use under the hood.
+            switch (key) {
+              case 'duplicate':
+                return 'This key is already in use. Keys must be unique.';
+              default:
+                return null;
+            }
+          }}
+        />
+        <NumberField
+          title="Inventory Quantity"
+          name="inventory"
+          isRequired={true}
+          value={this.props.formik.values.inventory}
+          onChange={this.props.formik.handleChange}
+          onBlur={this.props.formik.handleBlur}
+          isTouched={this.props.formik.touched.inventory}
+          errors={this.props.formik.errors.inventory}
+          renderError={key => {
+            switch (key) {
```

Comment:
Starting to like render callbacks here. At first I was sceptical and wanted s component but they‘re easier on the consumer.